### PR TITLE
Fix: report RV_EXCEPTION_NOT_THROWN ex created by factory method not thrown

### PR DESCRIPTION
**Pull request by JuditKnoll**
This PR fixes https://github.com/spotbugs/spotbugs/issues/2547.
In the current implementation RV_EXCEPTION_NOT_THROWN is only reported, when a constructor creates the exception, also the exception needs to have an `Error` or `Exception` suffix. 
This PR fixes both of these problems, by checking also those methods which returns an Exception, and the Exception is considered as an, when it has `Throwable` as one of it's superclasses.

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [X] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MethodReturnCheck.java [hunk 1]
```diff
@@ -308,18 +312,36 @@ private void sawMethodCallWithIgnoredReturnValue() {
...
+                }
             }
             state = State.SCAN;
         }
     }
 
+    private boolean isThrowableChild(@DottedClassName String className) {
+        try {
+            JavaClass[] superClasses = Repository.getSuperClasses(className);
+            return Arrays.stream(superClasses).anyMatch(clazz -> "java.lang.Throwable".equals(clazz.getClassName()));
```
[the remark this answers is not in the document]
Comment (JuditKnoll) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MethodReturnCheck.java]: Thanks, fixed. Great find, I will use this `Values` util class from now on. :)

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MethodReturnCheck.java [hunk 2]
```diff
@@ -308,18 +312,36 @@ private void sawMethodCallWithIgnoredReturnValue() {
...
+                    BugInstance warning = new BugInstance(this, "RV_EXCEPTION_NOT_THROWN", 1).addClassAndMethod(this).addMethod(callSeen)
+                            .describe(MethodAnnotation.METHOD_CALLED);
+                    bugAccumulator.accumulateBug(warning, SourceLineAnnotation.fromVisitedInstruction(this, callPC));
+                }
             }
             state = State.SCAN;
         }
     }
 
+    private boolean isThrowableChild(@DottedClassName String className) {
```
[the remark this answers is not in the document]
Comment (JuditKnoll) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MethodReturnCheck.java]: Thank you.
Looked at the code a bit more thoroughly, and found that `edu.umd.cs.findbugs.ba.ch.Subtypes2.instanceof()` is already used in this class, which almost does the same, but without requiring the `JavaClass` to be in the memory and it handles the `ClassNotFoundException`. So I decided to use this one instead.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MethodReturnCheck.java [hunk 3]
```diff
@@ -308,18 +312,36 @@ private void sawMethodCallWithIgnoredReturnValue() {
...
+                if (Const.CONSTRUCTOR_NAME.equals(callSeen.getName()) && isThrowableChild(callSeen.getClassName())) {
                     pattern = "RV_EXCEPTION_NOT_THROWN";
                 }
                 BugInstance warning = new BugInstance(this, pattern, priority).addClassAndMethod(this).addMethod(callSeen)
                         .describe(MethodAnnotation.METHOD_CALLED);
                 bugAccumulator.accumulateBug(warning, SourceLineAnnotation.fromVisitedInstruction(this, callPC));
+            } else {
+                String methodSig = callSeen.getMethodDescriptor().getSignature();
+                String returnTypeSig = methodSig.substring(methodSig.indexOf(')') + 1);
+                String returnType = ClassName.fromFieldSignature(returnTypeSig);
```
[the remark this answers is not in the document]
Comment (JuditKnoll) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MethodReturnCheck.java:319]: I couldn't find any way to get a `Method` type from `MethodDescriptor` (the `XMethod` type also wouldn't be much of a help), which would contain the info for the returntype, but found `SignatureParser.getReturnTypeSignature()`. This one basically does the same as I did, just hides it.

**Review by ThrawnCA — APPROVED** [review]

**Review by gtoison — APPROVED** [review]
Thanks for the changes!

**Pull request merged**